### PR TITLE
Add blacklister to repository management

### PR DESCRIPTION
**Pull request by blindpirate**
Fix https://github.com/gradle/gradle/issues/2853

Previous versions of Gradle would fall through to the next repository
if resolution in one repository failed. This may cause potentially
indeterministic resolution result. This PR changes this behaviour
and  will explicitly rethrow exceptions which occur in dependency
resolution instead of quietly continue to the next repository.

What's more, this PR introduces a RepositoryBlacklister. Exceptions
thrown during dependency resolution are categoried as follows:

- Caused by HTTP error status code (other than 2xx/3xx/404)
  - These exceptions would be considered as "recoverable" since the server seems still to be able to respond.
- Caused by other IOException/UncheckedIOException
  - These exceptions would be considered as "unrecoverable" and the repository would be blacklisted in the build.

This PR's target branch is set to #3041 intentionally to make it easier to review.

### CODE REVIEW

**Comment by bmuschko** [comment]
@blindpirate You might want to merge the HTTP timeout PR first. At the moment there's a lot of overlap between both PRs which is hard to review. I guess you merged the changes of the first PR into this one?

**Comment by blindpirate** [comment]
@bmuschko Now I've merged HTTP timeout branch into master and rebase master into this one so that it would be easy for you to review. @jjohannes thank you very much for taking over this!

**Comment by jjohannes** [comment]
@bmuschko can you have a final look at this?

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+            }
+        """
+
+        when:
+        fails('help')
+
+        then:
+        assertDependencyMetaDataReadTimeout(moduleA)
+
+        where:
```
Comment (bmuschko) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy]: This is missing some comments on the PR that succeeded this one e.g. https://github.com/gradle/gradle/pull/3041#discussion_r141990741. It had been fixed in the PR. I feel like there's still something wrong with what has or has not been merged. Why does this change even show up in this PR? It should already be in `release`.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+        moduleA.artifact.expectGet()
+        moduleB.pom.expectGet()
+        moduleB.artifact.expectGet()
+        succeeds('resolve', '--max-workers=1')
+
+        then:
+        downloadedLibsDir.assertContainsDescendants('a-1.0.jar', 'b-1.0.jar')
+    }
+
+    def "fails build if HTTP connection exceeds timeout when resolving metadata"() {
```
Comment (bmuschko) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy]: :x: We should try to unify the logic of these test cases. They are very similar. Maybe break out some fixture methods.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ConnectionFailureRepositoryBlacklister.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import com.google.common.collect.Sets;
+import org.apache.commons.lang.exception.ExceptionUtils;
+import org.gradle.api.UncheckedIOException;
+
+import java.io.IOException;
+import java.util.Set;
+
+public class ConnectionFailureRepositoryBlacklister implements RepositoryBlacklister {
+
+    private final Set<String> blacklistedRepositories = Sets.newConcurrentHashSet();
```
Comment (bmuschko) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ConnectionFailureRepositoryBlacklister.java]: :o: I am not sure if I interpret the Javadocs of [ConcurrentHashSet](http://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html?is-external=true) properly here but it sounds like you still need to synchronize.

> Retrieval operations (including get) generally do not block, so may overlap with update operations (including put and remove). Retrievals reflect the results of the most recently completed update operations holding upon their onset.

We should probably add some concurrency test for this class.

File: subprojects/docs/src/docs/release/notes.md [hunk 4]
```diff
@@ -272,6 +272,33 @@ There are better ways for re-using task logic, for example by using [task depend
...
+      > Could not get resource 'http://localhost:54347/repo/group/a/1.0/a-1.0.pom'.
+         > Could not GET 'http://localhost:54347/repo/group/a/1.0/a-1.0.pom'.
+            > Read timed out
+> Could not resolve group:b:1.0.
+  Required by:
+      project :
+   > Skipped due to earlier error
+```
+
+### Dependency resolution order
```
Comment (bmuschko) [inline subprojects/docs/src/docs/release/notes.md]: :x: I'd change this title to be more clear e.g. "Failure to resolve dependency from repository avoids checking other declared repositories".

File: subprojects/docs/src/docs/release/notes.md [hunk 5]
```diff
@@ -272,6 +272,33 @@ There are better ways for re-using task logic, for example by using [task depend
 
 The default output location in [EclipseClasspath](dsl/org.gradle.plugins.ide.eclipse.model.EclipseClasspath.html#org.gradle.plugins.ide.eclipse.model.EclipseClasspath:defaultOutputDir) changed from `${project.projectDir}/bin` to `${project.projectDir}/bin/default`.
 
+### Blacklist repository on `IOException`s
```
Comment (bmuschko) [inline subprojects/docs/src/docs/release/notes.md]: :x: I'd say it's rather a optimization than a breaking change. The breaking change is expressed by the next point. I'd list this under "new and noteworthy".

Needs to be more expressive. Maybe "Skip subsequent dependency request upon first failure to repository"?

File: subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/HttpErrorStatusCodeException.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.resource.transport.http;
+
+import org.gradle.internal.exceptions.Contextual;
+
+/**
+ * Signals that HTTP response has been received successfully but an error code is encountered (neither 2xx/3xx nor 404).
```
Comment (bmuschko) [inline subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/HttpErrorStatusCodeException.java]: :x: "neither 2xx/3xx nor 404" - That's incorrect based on the logic in `HttpClientHelper.wasSuccessful`.

`statusCode >= 200 && statusCode < 400`

File: subprojects/docs/src/docs/release/notes.md [hunk 7]
```diff
@@ -272,6 +272,33 @@ There are better ways for re-using task logic, for example by using [task depend
...
+  Required by:
+      project :
+   > Skipped due to earlier error
+```
+
+### Dependency resolution order
+
+Previous versions of Gradle would fall through to the next repository if resolution in one repository failed.
+This behaviour might cause potentially indeterministic resolution result. Now Gradle will explicitly rethrow exceptions which occur in dependency resolution instead of quietly continue to the next repository.
+This may break existed build. In this case, you should fix the error to avoid build indeterminism.
```
Comment (bmuschko) [inline subprojects/docs/src/docs/release/notes.md]: :x: "This behaviour might cause potentially indeterministic resolution result."

The common word would be "Nondeterministic".

"This may break existed build. In this case, you should fix the error to avoid build indeterminism."

I'd completely remove that sentence. This might just happen due to network issues.

File: subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/HttpErrorStatusCodeException.java (hunk 6 above)
Comment (oehme) [reply subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/HttpErrorStatusCodeException.java]: We check "wasMissing" before, so this is correct.

File: subprojects/docs/src/docs/release/notes.md [hunk 8]
```diff
@@ -272,6 +272,33 @@ There are better ways for re-using task logic, for example by using [task depend
 
 The default output location in [EclipseClasspath](dsl/org.gradle.plugins.ide.eclipse.model.EclipseClasspath.html#org.gradle.plugins.ide.eclipse.model.EclipseClasspath:defaultOutputDir) changed from `${project.projectDir}/bin` to `${project.projectDir}/bin/default`.
 
+### Blacklist repository on `IOException`s 
+
+In the event of a timeout (or any `IOException`s which are not caused by error HTTP status code), Gradle will skip subsequent connections to the same repository for the duration of the build. 
+The output of a build clearly indicates which request was skipped.
```
Comment (oehme) [inline subprojects/docs/src/docs/release/notes.md]: ⭕️ I'd remove this, it's too verbose for release notes.

File: subprojects/docs/src/docs/release/notes.md [hunk 9]
```diff
@@ -272,6 +272,33 @@ There are better ways for re-using task logic, for example by using [task depend
 
 The default output location in [EclipseClasspath](dsl/org.gradle.plugins.ide.eclipse.model.EclipseClasspath.html#org.gradle.plugins.ide.eclipse.model.EclipseClasspath:defaultOutputDir) changed from `${project.projectDir}/bin` to `${project.projectDir}/bin/default`.
 
+### Blacklist repository on `IOException`s 
+
+In the event of a timeout (or any `IOException`s which are not caused by error HTTP status code), Gradle will skip subsequent connections to the same repository for the duration of the build.
```
Comment (oehme) [inline subprojects/docs/src/docs/release/notes.md]: ⭕️ Let's just say "in case of repository failures". IOException is an implementation detail.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ConnectionFailureRepositoryBlacklister.java (hunk 3 above)
Comment (oehme) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ConnectionFailureRepositoryBlacklister.java]: What the JavaDoc says is that a `get` may not see a very recent `add`. Since we're only talking about a few CPU cycles here and the method is called from long-running operations (repository access), it is highly unlikely that this will ever happen. If it does happen, there is no loss of correctness, we just try to hit the broken repo one more time.

By synchronizing we would optimize for this rare corner case, but make the usual case slower.   I think the concurrent set is the right tool for the job here.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenDynamicResolveIntegrationTest.groovy [hunk 10]
```diff
@@ -376,24 +376,25 @@ Searched in the following locations:
         when:
         repo1.getModuleMetaData("group", "projectA").expectGet()
         projectA1.pom.expectGet()
-        projectA1.getArtifact().expectGet()
 
         repo2.getModuleMetaData("group", "projectA").expectGet()
         projectA2.pom.expectGetBroken()
```
Comment (oehme) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenDynamicResolveIntegrationTest.groovy]: ❌ Can you make sure we still have test coverage for the "missing" case? Otherwise we might be breaking behavior here.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainComponentMetaDataResolver.java [hunk 11]
```diff
@@ -79,66 +79,54 @@ public boolean isFetchingMetadataCheap(ComponentIdentifier identifier) {
...
-    private RepositoryChainModuleResolution findBestMatch(List<ComponentMetaDataResolveState> resolveStates, Collection<Throwable> failures) {
+    private RepositoryChainModuleResolution findBestMatch(List<ComponentMetaDataResolveState> resolveStates) throws Throwable {
         LinkedList<ComponentMetaDataResolveState> queue = new LinkedList<ComponentMetaDataResolveState>();
         queue.addAll(resolveStates);
 
         LinkedList<ComponentMetaDataResolveState> missing = new LinkedList<ComponentMetaDataResolveState>();
 
         // A first pass to do local resolves only
-        RepositoryChainModuleResolution best = findBestMatch(queue, failures, missing);
+        RepositoryChainModuleResolution best = findBestMatch(queue, missing);
```
Comment (oehme) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainComponentMetaDataResolver.java]: ❌ What if the local repo has a problem (e.g. a broken maven-metadata.xml), but the remote one is fine?

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ConnectionFailureRepositoryBlacklisterTest.groovy [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+import spock.lang.Subject
+import spock.lang.Unroll
+
+class ConnectionFailureRepositoryBlacklisterTest extends Specification {
+
+    @Subject RepositoryBlacklister blacklister = new ConnectionFailureRepositoryBlacklister()
+
+    def "initializes with no blacklisted repositories"() {
+        expect:
+        blacklister.blacklistedRepositories.empty
```
Comment (oehme) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ConnectionFailureRepositoryBlacklisterTest.groovy]: ⭕️ This is testing internal implementation details. As far as I can see we are only exposing the `Set` to test it here. I'd remove this and only test through the API we actually care about: `blackList()` and `isBlackListed()`

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepositoryTest.groovy [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+import org.gradle.internal.resolve.ModuleVersionResolveException
+import org.gradle.internal.resolve.result.BuildableArtifactResolveResult
+import org.gradle.internal.resolve.result.BuildableArtifactSetResolveResult
+import org.gradle.internal.resolve.result.BuildableComponentArtifactsResolveResult
+import org.gradle.internal.resolve.result.BuildableModuleComponentMetaDataResolveResult
+import org.gradle.internal.resolve.result.BuildableModuleVersionListingResolveResult
+import spock.lang.Specification
+import spock.lang.Subject
+
+class ErrorHandlingModuleComponentRepositoryTest extends Specification {
```
Comment (oehme) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepositoryTest.groovy]: ❌ This test is over-asserting so much that it basically repeats the implementation line by line. It also tests many interactions in every method. This will break on all kinds of internal refactorings that don't change behavior at all.

A unit test should only test one interaction at a time. For everything that the test method does not care about, a stub should be used instead of a mock.

E.g. there should be one test asserting that the repository access blacklists a repo when it threw an exception. That test should assert nothing else.

For simple types like our BlackLister it is probably easier to use state-based testing, since we already have the appropriate query methods. I.e. use the real implementation and its query method instead of a mock.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ConnectionFailureRepositoryBlacklister.java (hunk 3 above)
Comment (blindpirate) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ConnectionFailureRepositoryBlacklister.java]: Things are much worse than the incorrectness. `HashSet` and `HashMap` are not thread-safe and concurrent write operation might cause infinite loops in very rare cases. Actually I had met 100% CPU usage several times previously.  https://bugs.openjdk.java.net/browse/JDK-6423457

File: subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/HttpErrorStatusCodeException.java (hunk 6 above)
Comment (blindpirate) [reply subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/HttpErrorStatusCodeException.java]: Thanks for pointing out. The 404 check is here: https://github.com/gradle/gradle/blob/5c75c8c9be85181db1145c810b33b7ead54562a1/subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/HttpClientHelper.java#L152

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy (hunk 1 above)
Comment (blindpirate) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy]: Good catch, I did miss something when merging and I've fixed this. The reason why this test shows up here is this file was renamed from 
 `DependencyResolveTimeoutIntegrationTest` to `DependencyUnresolvedModuleIntegrationTest` in https://github.com/gradle/gradle/commit/65c833548f8fb3d3c5097cd3db8545979aaab43c

File: subprojects/docs/src/docs/release/notes.md (hunk 5 above)
Comment (blindpirate) [reply subprojects/docs/src/docs/release/notes.md]: Sure, moved.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy (hunk 2 above)
Comment (blindpirate) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy]: Thanks for advice, I've compressed them, would you like to take a look?

File: subprojects/docs/src/docs/release/notes.md (hunk 4 above)
Comment (blindpirate) [reply subprojects/docs/src/docs/release/notes.md]: Changed to `Avoid checking other repositories when dependency resolution in one repository fails`, would it be better?

File: subprojects/docs/src/docs/release/notes.md (hunk 7 above)
Comment (blindpirate) [reply subprojects/docs/src/docs/release/notes.md]: Done.

File: subprojects/docs/src/docs/release/notes.md (hunk 9 above)
Comment (blindpirate) [reply subprojects/docs/src/docs/release/notes.md]: Done.

File: subprojects/docs/src/docs/release/notes.md (hunk 8 above)
Comment (blindpirate) [reply subprojects/docs/src/docs/release/notes.md]: Done.

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ConnectionFailureRepositoryBlacklisterTest.groovy (hunk 12 above)
Comment (blindpirate) [reply subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ConnectionFailureRepositoryBlacklisterTest.groovy]: Done.

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepositoryTest.groovy (hunk 13 above)
Comment (blindpirate) [reply subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepositoryTest.groovy]: I've broken it into small piece code of `when-then` blocks here: https://github.com/gradle/gradle/blob/79dd5b101a208061bfd43d8620d81fc70d94027d/subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepositoryTest.groovy

Hope this would be better.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy (hunk 1 above)
Comment (jjohannes) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy]: Maybe I also mixed something when I rebased this PR. `DependencyResolveTimeoutIntegrationTest` is not needed anymore then, because everything is in `DependencyUnresolvedModuleIntegrationTest` now, right? Can you remove `DependencyResolveTimeoutIntegrationTest`?

Comment (blindpirate) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy]: Oh yes, sure.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenDynamicResolveIntegrationTest.groovy (hunk 10 above)
Comment (jjohannes) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenDynamicResolveIntegrationTest.groovy]: @blindpirate I think we should add a test like this:

```
    def "dynamic version ignores missing module in one repository when available in another repository"()  {
        given:
        def repo1 = mavenHttpRepo("repo1")
        def repo2 = mavenHttpRepo("repo2")

        def projectA2 = repo2.module('group', 'projectA', '1.5').publish()

        buildFile << """
        repositories {
            maven { url '${repo1.uri}' }
            maven { url '${repo2.uri}' }
        }
        configurations { compile }
        dependencies {
            compile 'group:projectA:1.+'
        }

        task retrieve(type: Sync) {
            into 'libs'
            from configurations.compile
        }
        """

        when:
        repo1.getModuleMetaData("group", "projectA").expectGetMissing()
        repo1.directory("group", "projectA").expectGetMissing()
        
        repo2.getModuleMetaData("group", "projectA").expectGet()
        projectA2.pom.expectGet()
        projectA2.artifact.expectGet()

        then:
        succeeds 'retrieve'
    }
```

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainComponentMetaDataResolver.java (hunk 11 above)
Comment (jjohannes) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainComponentMetaDataResolver.java]: Can you check if we have a test for this behavior? It is ok if it fails, but it should be like this in the current version already. We need to make sure that we do not change the behavior.

Comment (blindpirate) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainComponentMetaDataResolver.java]: @jjohannes Done here:https://github.com/gradle/gradle/blob/21712f94de5cd0096b0c55ece8bac8cd7bd68af8/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenLocalRepoResolveIntegrationTest.groovy#L127

I'm not sure if this is correct, could a repository have both local access and remote access at the same time?

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenDynamicResolveIntegrationTest.groovy [hunk 14]
```diff
@@ -350,52 +298,80 @@ Searched in the following locations:
...
         and:
-        run 'retrieve'
+        succeeds 'retrieve'
+
+        then:
+        file('libs').assertHasDescendants('projectA-1.5.jar')
+    }
+
+    def "dynamic version ignores missing module in one repository when available in another repository"() {
+        given:
```
Comment (blindpirate) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenDynamicResolveIntegrationTest.groovy]: @jjohannes Done here.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainComponentMetaDataResolver.java (hunk 11 above)
Comment (jjohannes) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainComponentMetaDataResolver.java]: The test is ok. We always first try the local access and then the remote access (if local is missing) as far as I understand. On error in local, we exit directly (without trying remote).
`ComponentMetaDataResolveState` always first tries local access and if it fails with an error, the complete _resolve_ operation fails. The test above however fails before we even get there.

Another test that exercises this code path would be good as well. For that situation, the local metadata needs to be corrupt. We could do a similar test as above:

```
      given:
        def corruptLocalPom = m2.mavenRepo().rootDir.file('group/projectA/1.2/projectA-1.2.pom')
        corruptLocalPom << "invalid content"

        if (remoteStatus == 'valid') {
            def remoteRepo = maven("remote")
            remoteRepo.module('group', 'projectA', '1.2').publish()

            buildFile << """
            repositories{
                maven { url "${remoteRepo.uri}" }
            }
            """
        }

        when:
        runAndFail 'retrieve'

        then:
        failure.assertHasCause("Could not parse POM $corruptLocalPom.absolutePath")

        where:
        remoteStatus << ['valid', 'invalid']
```

:x: In fact, it looks like this test fails without the change. We might need to rethink the change with the error handling. 🙁

**Comment by oehme** [comment]
I would say keep the behavior as it was before: If local fails, still try remote. We can change that independently later.

**Comment by blindpirate** [comment]
I think we should do something similar: we try remote only when we know "it's not in local repository".

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainComponentMetaDataResolver.java (hunk 11 above)
Comment (blindpirate) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainComponentMetaDataResolver.java]: Looks like it's not mocking a failure in local access here because `m2.mavenRepo().rootDir.file('group/projectA/1.2/projectA-1.2.pom')` is `mavenLocal()`, so actually there're 2 repos here: `mavenLocal()` and `remoteRepo`.

This test is testing fall through next repository.

**Comment by blindpirate** [comment]
@oehme 
Looks like a remote repository's local access is NOT `mavenLocal()`. Actually, a remote repository's `local access` doesn't do anything: https://github.com/gradle/gradle/blob/bf784b01afb4f155a0b41adaa77995dd38f301d7/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/ExternalResourceResolver.java#L461

So a corrupted `maven-metadata.xml` on disk will cause a failure in `mavenLocal()` repository, it has nothing to do with `local access`.

@jjohannes This could explain why the test case you gave failed without the change, it's a fall through to the next repository.

Did I miss anything?

**Comment by oehme** [comment]
@blindpirate I assumed that the "local access" is a look-up in Gradle's cache. But maybe that's handled somewhere else. I'll let Jendrik and you decide on this issue.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainComponentMetaDataResolver.java (hunk 11 above)
Comment (jjohannes) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainComponentMetaDataResolver.java]: Yes you are right. This change is expected. Sorry for the confusion.

**Comment by jjohannes** [comment]
The `RepositoryAccess` is a chain, which will ask Gradle caches for results, before it even gets to the  maven repository local access (which does nothing). The code structure is rather confusing and it is very hard to follow what _local access_ actually means. (Here in fact it means nothing - the repository access chain will take the result from a cache or will get it remotely)

The only, unlikely, problem I see is that a cache that was persisted to disk is used and somehow is broken and produces an error. But even if that can happen at all, I don't think this is the place where such an error should get special treatment.

I approve this PR for 4.3.
